src/TorchSharp/DisposeScope.cs

```diff
@@ -0,0 +1,207 @@
+        }
+
+        public (T1 first, T2 second, T3 third) ExcludeGlobally<T1, T2, T3>(T1 first, T2 second, T3 third)
+            where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+        {
+            Detach(new IDisposable[] { first, second, third });
+            return (first, second, third);
+        }
+
+        public void Detach(IDisposable[] disposables) => Detach((IEnumerable<IDisposable>)disposables);
+
+        public void Detach(IEnumerable<IDisposable> disposables)
+        {
+            foreach (var disposable in disposables) {
+                if (Disposables.Remove(disposable)) {
+                    if (disposable is torch.Tensor tensor) {
+                        tensor.OwningDisposeScope = null;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Disposes everything currenly in the dispose scope.
+        /// </summary>
+        public void DisposeEverything() => DisposeEverythingBut(Enumerable.Empty<IDisposable>());
+
+        /// <summary>
+        /// As an intermediate step, you can dispose all the tensors/disposables currently scheduled for dispose, to
+        /// clear up some memory without creating a new scope. Note that this doesn't permanently exclude the
+        /// tensors from disposing, use Exclude for that. Also, excluded tensors don't need to be included
+        /// here.
+        /// </summary>
+        public void DisposeEverythingBut(IEnumerable<IDisposable> keep)
+        {
+            var oldList = Disposables;
+            Disposables = keep.ToHashSet(ReferenceEqualityComparer<IDisposable>.Default);
+            foreach (var disposable in oldList) {
+                if (!Disposables.Contains(disposable)) {
+                    disposable.Dispose();
+                }
+            }
+        }
+
+        public T DisposeEverythingBut<T>(T keep) where T : IDisposable
+        {
+            DisposeEverythingBut(new IDisposable[] { keep });
+            return keep;
+        }
+
+        public (T1 first, T2 second) DisposeEverythingBut<T1, T2>(T1 first, T2 second)
+            where T1 : IDisposable where T2 : IDisposable
+        {
+            DisposeEverythingBut(new IDisposable[] { first, second });
+            return (first, second);
+        }
+
+        public (T1 first, T2 second, T3 third) DisposeEverythingBut<T1, T2, T3>(T1 first, T2 second, T3 third)
+            where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+        {
+            DisposeEverythingBut(new IDisposable[] { first, second, third });
+            return (first, second, third);
+        }
+
+        public void Dispose()
+        {
+            foreach (var disposable in Disposables.ToList()) {
+                disposable.Dispose();
+            }
+
+            Disposables.Clear();
+            _disposeScopeManager.RemoveDisposeScope(this);
+        }
+
+        public void WasDisposed(IDisposable disposable)
```

Comment:
Not sure I agree, it's for the Disposable to notify the DisposeScope that it was disposed so the DisposeScope can stop tracking it. It's not the same as detach, because for book-keeping purposes, we want to keep track of the fact that it was actually disposed.

If someone creates another use for this system - keeping track of training batches for instance - they may want the same integration opportunities?

Reply:
Okay. With the summary, the name makes sense. 

Looking at the code, I have another question: If I'm correct, when the call to Dispose() on line 247 gets to the tensor, it will call 'WasDisposed' on, which will remove it from the list individually, rather than relying on clearing it on line 250. 

Maybe the tensor should have its OwningDisposeScope property cleared before the call to Dispose() on line 247, to avoid the call to WasDisposed(). That seems to deal more efficiently with the wholesale disposal of all tensors in the scope.

Something like:

```C#
            foreach (var disposable in Disposables.ToList()) {
                if (disposable is torch.Tensor tensor) tensor.OwningDisposeScope = null;
                disposable.Dispose();
            }

            _disposeScopeManager.DisposedInScopeCount += Disposables.Count;
            Disposables.Clear();
            _disposeScopeManager.RemoveDisposeScope(this);
```

Sorry for being so nit-picky, but this is going to be used all the time, and on a lot of critical performance paths, so I want to make sure we squeeze out every little bit of performance.

Reply:
Yes, it gets the run-around. I'll change it to short circuit the call. It's not that expensive as it is, it's just a single method call, but it's cleaner.

Reply:
fixed

Reply:
NIT: `MarkDisposed`?